### PR TITLE
add proxy support, add new params to readme documentation

### PR DESCRIPTION
**Pull request by terrimonster**
Added proxy support to types and providers, as well as unit tests. I spun up a squid vagrant instance and a base instance, ran the tests/archive.pp with the new params set, and all ran well. Added the params to the readme documentation as well.

### CODE REVIEW

**Comment by nanliu** [comment]
Could we just infer the proxy type from the proxy url?

**Comment by terrimonster** [comment]
We probably could. I went the lazy route of leaving that up to the end user. I could code it in, however, if that is preferred.

**Comment by nanliu** [comment]
Yeah, let's just infer it if it's not a big change. Also for the parameter, should we just simplify to 'proxy', any suggestions?

**Comment by terrimonster** [comment]
That all can be done pretty easily. Give me a few.

**Comment by jyaworski** [comment]
I think this was merged around the same time that #82 was. Can you rebase?

**Comment by terrimonster** [comment]
Rebased, style-cleaned. I have kept in the proxy type and server. After thinking abotu this and discussions, it is risky to make assumptions and take the control out of the user's hands.

**Comment by jyaworski** [comment]
Looks good. Can we rebase this down to one commit?

File: lib/puppet/type/archive.rb [hunk 1]
```diff
@@ -150,6 +150,16 @@ def should_to_s(value)
     desc 'extract command group (using this option will configure the archive file permisison to 0644 so the user can read the file).'
   end
 
+  newparam(:proxy_type) do
+    desc 'proxy type (none|ftp|http|https)'
+    newvalues(:none, :ftp, :http, :https)
+    defaultto(:none)
```
Comment (nanliu) [inline lib/puppet/type/archive.rb]: We can defaultto a sensible value based on proxy_server and let user override. Let me find you an example.

File: lib/puppet/type/archive.rb [hunk 2]
```diff
@@ -150,6 +150,16 @@ def should_to_s(value)
...
+  end
+
+  newparam(:proxy_server) do
+    desc 'proxy address to use when accessing source'
+  end
+
```
Comment (nanliu) [inline lib/puppet/type/archive.rb]: I think if you leave :proxy_type without any default you can at the end here do:
```ruby
validate
  self[:proxy_type] ||= URI(self[:proxy_server]).scheme if self[:proxy_server]
end
```

This should make it more convenient than forcing user to specify proxy_type every time.

**Comment by terrimonster** [comment]
Added your suggestion, Nan. Thank you very much for the code block :)

**Comment by jyaworski** [comment]
This looks great.

File: lib/puppet/type/archive.rb (hunk 1 above)
Comment (nanliu) [inline lib/puppet/type/archive.rb]: FYI with the new code please remove this defaultto

**Comment by nanliu** [comment]
Ok, please remove the defaultto and squash commit, then we are ready to merge.

File: spec/unit/puppet/provider/archive/wget_spec.rb [hunk 3]
```diff
@@ -78,5 +78,21 @@
         provider.download(name)
       end
     end
+
+    context 'proxy specified' do
+      let(:resource_properties) do
+        {
+          :name => name,
+          :source => 'http://home.lan/example.zip',
+          :proxy_type => 'https',
```
Comment (nanliu) [inline spec/unit/puppet/provider/archive/wget_spec.rb]: Also with the new code we can rewrite this test without providing :proxy_type and see if it just do the right thing.

**Comment by nanliu** [comment]
Fix #80.

**Comment by nanliu** [comment]
Bla meant to merge and close.

File: lib/puppet/type/archive.rb (hunk 2 above)
Comment (alexjfisher) [inline lib/puppet/type/archive.rb]: @nanliu But is this correct?

`proxy_type` is used to set an environment variable such as `http_proxy` or `https_proxy`.  It should default to the scheme of `source` not the scheme of `proxy_server`?

Comment (nanliu) [inline lib/puppet/type/archive.rb]: @alexjfisher, I believe so.

We are assuming if `proxy_server=https://foo_proxy` we are using `https` proxy_type, and we don't care if you are trying to download a file from `http://somefile`. Also the user can be explicit about the proxy_type and we won't guess.

Comment (alexjfisher) [inline lib/puppet/type/archive.rb]: Still not sure I'm convinced.  But yeah, you can always override proxy_type. You may have guessed, but I was trying to debug https://github.com/voxpupuli/puppet-archive/issues/277

with wget on the command line for example
`http_proxy=http://proxy.example.com:3128 wget https://www.example.com/somefile`
the proxy in the environment variable `http_proxy` would not be used.

You'd have to do
`https_proxy=http://proxy.example.com:3128 wget https://www.example.com/somefile`
instead and the proxy/squid would tunnel the https request.  (Having it intercept https can be done but is *far* less common I think)

**Pull request merged**